File: src/test/java/com/kinetica/spark/GPUdbUtil.java
package com.kinetica.spark;

import java.net.InetAddress;
import java.net.NetworkInterface;
import java.net.SocketException;
import java.util.Enumeration;

import com.gpudb.GPUdb;
import com.gpudb.GPUdbException;
import com.gpudb.RecordObject;
import com.gpudb.protocol.CreateTableRequest;

import org.apache.hadoop.conf.Configuration;


/**
 * Provides common functions required for Spark-based ingestion:
 * 
 * <ul>
 * <li>getGPUdb - returns a GPUdb connection based on Spark configuration</li>
 * <li>getLocalIP - returns local interface IP with the given IP prefix</li>
 * <li>createTable - creates a table with the given GPUdb connection or GPUdb
 *     URL, if it doesn't already exist</li>
 * </ul>
 */
public class GPUdbUtil
{
	private static final String CONFIG_GPUDB_HOST = "gpudb.host";
	private static final String CONFIG_GPUDB_PORT = "gpudb.port";

	/**
	 * Creates a GPUdb connection, using the host and port specified within the
	 * given Hadoop configuration.  The configuration must contain the following
	 * properties:
	 * <ul>
	 * <li>gpudb.host</li>
	 * <li>gpudb.port</li>
	 * </ul>
	 * 
	 * @param config Hadoop configuration, containing connection properties
	 * @return GPUdb connection created
	 * @throws GPUdbException if an error occurs creating the GPUdb connection
	 */
	public static GPUdb getGPUdb(Configuration config) throws GPUdbException
	{
		String gpudbUrl = "http://" + config.get(CONFIG_GPUDB_HOST) + ":" + config.get(CONFIG_GPUDB_PORT);
		GPUdb gpudb = new GPUdb(gpudbUrl);
		return gpudb;
	}

	/**
	 * Searches the current host's interfaces for the first one with an IP
	 * address matching the given prefix
	 * 
	 * @param localHostIpPrefix prefix used to match against local interface IPs
	 * @return the first local interface IP matching the given prefix; null, if
	 *         no match is found
	 * @throws SocketException if an error occurs searching local interfaces
	 */
	public static String getLocalIP(String localHostIpPrefix) throws SocketException
	{
		String localHostIPAddress = null;
	
		Enumeration<NetworkInterface> e = NetworkInterface.getNetworkInterfaces();
		while(e.hasMoreElements())
		{
			NetworkInterface n = e.nextElement();
			Enumeration<InetAddress> ee = n.getInetAddresses();
			while (ee.hasMoreElements())
			{
				InetAddress i = ee.nextElement();
				if (i.getHostAddress().startsWith(localHostIpPrefix))
					localHostIPAddress = i.getHostAddress();
			}
		}
		
		return localHostIPAddress;
	}

	/**
	 * Creates a table within GPUdb
	 * 
	 * @param gpudbUrl HTTP address of GPUdb server in which table should be
	 *        created
	 * @param collectionName name of collection in which to create table
	 * @param tableName name of table to create
	 * @param tableType class name of table type to create
	 * @return true, if table was created successfully;
	 *         false, if not because table already exists
	 * @throws GPUdbException if table didn't exist and failed to be created
	 */
	public static boolean createTable(String gpudbUrl, String collectionName, String tableName, Class<? extends RecordObject> tableType) throws GPUdbException
	{
		GPUdb gpudb = new GPUdb(gpudbUrl);
		
		return createTable(gpudb, collectionName, tableName, tableType);
	}

	/**
	 * Creates a table within GPUdb
	 * 
	 * @param gpudb GPUdb connection to use to create table
	 * @param collectionName name of collection in which to create table
	 * @param tableName name of table to create
	 * @param tableType class name of table type to create
	 * @return true, if table was created successfully;
	 *         false, if not because table already exists
	 * @throws GPUdbException if table didn't exist and failed to be created
	 */
	public static boolean createTable(GPUdb gpudb, String collectionName, String tableName, Class<? extends RecordObject> tableType) throws GPUdbException
	{
		boolean alreadyExists = false;

    	try
		{
		    String typeId = RecordObject.createType(tableType, gpudb);
		    gpudb.addKnownType(typeId, tableType);
		    gpudb.createTable(tableName, typeId, GPUdb.options(CreateTableRequest.Options.COLLECTION_NAME, collectionName));
		}
		catch (GPUdbException e)
		{
			if (e.getMessage().contains("already exists"))
				alreadyExists = true;
			else
			    throw e;
		}
    	
    	return !alreadyExists;
	}
}

File: src/test/scala/com/kinetica/spark/streaming/GPUdbUtil.scala
package com.kinetica.spark.streaming

import java.net.InetAddress
import java.net.NetworkInterface
import java.net.SocketException
import java.util.Enumeration
import com.gpudb._
import com.gpudb.protocol.CreateTableRequest
import org.apache.hadoop.conf.Configuration

//remove if not needed
import scala.collection.JavaConversions._

object GPUdbUtil {

    /**
     * Creates a table within GPUdb
     *
     * @param gpudbUrl HTTP address of GPUdb server in which table should be
     *        created
     * @param collectionName name of collection in which to create table
     * @param tableName name of table to create
     * @param tableType class name of table type to create
     * @return true, if table was created successfully;
     *         false, if not because table already exists
     * @throws GPUdbException if table didn't exist and failed to be created
     */
    def createTable(
        gpudbUrl: String,
        collectionName: String,
        tableName: String,
        tableType: Class[_ <: RecordObject]): Boolean = {
        val gpudb: GPUdb = new GPUdb(gpudbUrl)
        createTable(gpudb, collectionName, tableName, tableType)
    }

    /**
     * Creates a table within GPUdb
     *
     * @param gpudb GPUdb connection to use to create table
     * @param collectionName name of collection in which to create table
     * @param tableName name of table to create
     * @param tableType class name of table type to create
     * @return true, if table was created successfully;
     *         false, if not because table already exists
     * @throws GPUdbException if table didn't exist and failed to be created
     */
    def createTable(
        gpudb: GPUdb,
        collectionName: String,
        tableName: String,
        tableType: Class[_ <: RecordObject]): Boolean = {
        var alreadyExists: Boolean = false
        try {
            val typeId: String = RecordObject.createType(tableType, gpudb)
            gpudb.addKnownType(typeId, tableType)
            gpudb.createTable(
                tableName,
                typeId,
                GPUdbBase.options(
                    CreateTableRequest.Options.COLLECTION_NAME,
                    collectionName))
        } catch {
            case e: GPUdbException =>
                if (e.getMessage.contains("already exists")) alreadyExists = true
                else throw e

        }
        !alreadyExists
    }
    
    def getData(myType:Type) : Record = {
        val gr = myType.newInstance();
        val columnCount = myType.getColumnCount
        val r = scala.util.Random
        for (column <- myType.getColumns) { 
            
            //println(" &&&&&&& " + column.getProperties.mkString)
            if (column.getType().toString().contains("java.lang.Double")) {
                gr.put(column.getName(), r.nextDouble())
            } else if (column.getType().toString().contains("java.lang.Float")) {
                gr.put(column.getName(), r.nextFloat())
            } else if (column.getType().toString().contains("java.lang.Integer")) {
                if ( column.getProperties.mkString.toLowerCase.contains("int16") ) {
                    gr.put(column.getName(), r.nextInt(2^16))
                } else if ( column.getProperties.mkString.toLowerCase.contains("int8") ) { 
                    gr.put(column.getName(), r.nextInt(2^8))
                } else gr.put(column.getName(), r.nextInt())
            } else if (column.getType().toString().contains("java.lang.Long")) {
                if ( column.getProperties.mkString.toLowerCase.contains("timestamp") ) {
                    var ts = r.nextLong()
                    ts = if (ts < -30610224000000L) -30610224000000L else ts
                    ts = if (ts > 29379542399999L) 29379542399999L else ts
                    gr.put(column.getName(), ts);
                } else {
                    gr.put(column.getName(), r.nextLong())
                }
            } else {
                if ( column.getProperties.mkString.toLowerCase.contains("date") )
                    gr.put(column.getName(), "1963-12-25");
                else if ( column.getProperties.mkString.toLowerCase.contains("ipv4") )
                    gr.put(column.getName(), "172.12.24.63");
                else if ( column.getProperties.mkString.toLowerCase.contains("decimal") )
                    gr.put(column.getName(), "155.52");
                else gr.put(column.getName(), scala.util.Random.alphanumeric.take(1).mkString);
            }
        }
        gr
    }

}
